File: python/mirage/workspace/executor/builtins/table.py
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.
# ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

from mirage.shell.types import ShellBuiltin as SB
from mirage.workspace.executor.builtins.alias.alias import (alias_builtin,
                                                            unalias_builtin)
from mirage.workspace.executor.builtins.command.command import command_builtin
from mirage.workspace.executor.builtins.condition.condition import test_builtin
from mirage.workspace.executor.builtins.control.control import (
    break_builtin, colon_builtin, continue_builtin, exit_builtin,
    false_builtin, return_builtin, true_builtin)
from mirage.workspace.executor.builtins.declare.export import export_builtin
from mirage.workspace.executor.builtins.declare.local import local_builtin
from mirage.workspace.executor.builtins.dirs.cd import cd_builtin
from mirage.workspace.executor.builtins.dirs.pwd import pwd_builtin
from mirage.workspace.executor.builtins.echo.echo import echo_builtin
from mirage.workspace.executor.builtins.env.env import env_builtin
from mirage.workspace.executor.builtins.eval.eval import eval_builtin
from mirage.workspace.executor.builtins.exec.exec import exec_builtin
from mirage.workspace.executor.builtins.getopts.getopts import getopts_builtin
from mirage.workspace.executor.builtins.history.history import history_builtin
from mirage.workspace.executor.builtins.let.let import let_builtin
from mirage.workspace.executor.builtins.lookup.lookup import (type_builtin,
                                                              which_builtin)
from mirage.workspace.executor.builtins.man.man import man_builtin
from mirage.workspace.executor.builtins.mapfile.mapfile import mapfile_builtin
from mirage.workspace.executor.builtins.printenv.printenv import \
    printenv_builtin
from mirage.workspace.executor.builtins.printf.printf import printf_builtin
from mirage.workspace.executor.builtins.read.read import read_builtin
from mirage.workspace.executor.builtins.script.bash import bash_builtin
from mirage.workspace.executor.builtins.script.source import source_builtin
from mirage.workspace.executor.builtins.set.set import set_builtin
from mirage.workspace.executor.builtins.shift.shift import shift_builtin
from mirage.workspace.executor.builtins.shopt.shopt import shopt_builtin
from mirage.workspace.executor.builtins.sleep.sleep import sleep_builtin
from mirage.workspace.executor.builtins.timeout.timeout import timeout_builtin
from mirage.workspace.executor.builtins.trap.trap import trap_builtin
from mirage.workspace.executor.builtins.types import BuiltinFn
from mirage.workspace.executor.builtins.umask.umask import umask_builtin
from mirage.workspace.executor.builtins.unset.unset import unset_builtin
from mirage.workspace.executor.builtins.whoami.whoami import whoami_builtin
from mirage.workspace.executor.builtins.xargs.xargs import xargs_builtin

# The one map from an executor-run builtin word to its handler; the
# dispatcher does a single lookup here instead of one arm per word. Two
# ShellBuiltin groups are deliberately absent: the job builtins (wait,
# fg, kill, jobs, disown, ps) route through JOB_HANDLERS in
# executor/command/command.py because they need the job table, and
# python/python3/node/js are general mount commands
# (commands/builtin/general), reserved in ShellBuiltin only so no CLI
# can take the name. tests/workspace/executor/builtins/test_table.py
# pins that everything else is here.
BUILTINS: dict[str, BuiltinFn] = {
    SB.PWD: pwd_builtin,
    SB.CD: cd_builtin,
    SB.EXPORT: export_builtin,
    SB.UNSET: unset_builtin,
    SB.LOCAL: local_builtin,
    SB.SET: set_builtin,
    SB.READ: read_builtin,
    SB.MAPFILE: mapfile_builtin,
    SB.READARRAY: mapfile_builtin,
    SB.SHIFT: shift_builtin,
    SB.GETOPTS: getopts_builtin,
    SB.LET: let_builtin,
    SB.TRAP: trap_builtin,
    SB.SHOPT: shopt_builtin,
    SB.UMASK: umask_builtin,
    SB.ALIAS: alias_builtin,
    SB.UNALIAS: unalias_builtin,
    SB.EXEC: exec_builtin,
    SB.TEST: test_builtin,
    SB.BRACKET: test_builtin,
    SB.DOUBLE_BRACKET: test_builtin,
    SB.ECHO: echo_builtin,
    SB.PRINTF: printf_builtin,
    SB.SOURCE: source_builtin,
    SB.DOT: source_builtin,
    SB.EVAL: eval_builtin,
    SB.COMMAND: command_builtin,
    SB.TYPE: type_builtin,
    SB.WHICH: which_builtin,
    SB.TRUE: true_builtin,
    SB.FALSE: false_builtin,
    SB.COLON: colon_builtin,
    SB.BREAK: break_builtin,
    SB.CONTINUE: continue_builtin,
    SB.RETURN: return_builtin,
    SB.EXIT: exit_builtin,
    SB.PRINTENV: printenv_builtin,
    SB.ENV: env_builtin,
    SB.WHOAMI: whoami_builtin,
    SB.MAN: man_builtin,
    SB.HISTORY: history_builtin,
    SB.SLEEP: sleep_builtin,
    SB.BASH: bash_builtin,
    SB.SH: bash_builtin,
    SB.XARGS: xargs_builtin,
    SB.TIMEOUT: timeout_builtin,
}

File: typescript/packages/core/src/workspace/executor/builtins/table.ts
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.
// ========= Copyright 2026 @ Strukto.AI All Rights Reserved. =========

import { ShellBuiltin as SB } from '../../../shell/types.ts'
import { aliasBuiltin, unaliasBuiltin } from './alias/alias.ts'
import { commandBuiltin } from './command/command.ts'
import { testBuiltin } from './condition/condition.ts'
import {
  breakBuiltin,
  colonBuiltin,
  continueBuiltin,
  exitBuiltin,
  falseBuiltin,
  returnBuiltin,
  trueBuiltin,
} from './control/control.ts'
import { exportBuiltin } from './declare/export.ts'
import { localBuiltin } from './declare/local.ts'
import { cdBuiltin } from './dirs/cd.ts'
import { pwdBuiltin } from './dirs/pwd.ts'
import { echoBuiltin } from './echo/echo.ts'
import { envBuiltin } from './env/env.ts'
import { evalBuiltin } from './eval/eval.ts'
import { execBuiltin } from './exec/exec.ts'
import { getoptsBuiltin } from './getopts/getopts.ts'
import { historyBuiltin } from './history/history.ts'
import { letBuiltin } from './let/let.ts'
import { typeBuiltin, whichBuiltin } from './lookup/lookup.ts'
import { manBuiltin } from './man/man.ts'
import { mapfileBuiltin } from './mapfile/mapfile.ts'
import { printenvBuiltin } from './printenv/printenv.ts'
import { printfBuiltin } from './printf/printf.ts'
import { readBuiltin } from './read/read.ts'
import { bashBuiltin } from './script/bash.ts'
import { sourceBuiltin } from './script/source.ts'
import { setBuiltin } from './set/set.ts'
import { shiftBuiltin } from './shift/shift.ts'
import { shoptBuiltin } from './shopt/shopt.ts'
import { sleepBuiltin } from './sleep/sleep.ts'
import { timeoutBuiltin } from './timeout/timeout.ts'
import { trapBuiltin } from './trap/trap.ts'
import type { BuiltinFn } from './types.ts'
import { umaskBuiltin } from './umask/umask.ts'
import { unsetBuiltin } from './unset/unset.ts'
import { whoamiBuiltin } from './whoami/whoami.ts'
import { xargsBuiltin } from './xargs/xargs.ts'

// The one map from an executor-run builtin word to its handler; the
// dispatcher does a single lookup here instead of one arm per word. Two
// ShellBuiltin groups are deliberately absent: the job builtins (wait, fg,
// kill, jobs, disown, ps) route through JOB_HANDLERS in
// executor/command.ts because they need the job table, and
// python/python3/node/js are general mount commands
// (commands/builtin/general), reserved in ShellBuiltin only so no CLI can
// take the name. table.test.ts pins that everything else is here.
export const BUILTINS: ReadonlyMap<string, BuiltinFn> = new Map<string, BuiltinFn>([
  [SB.PWD, pwdBuiltin],
  [SB.CD, cdBuiltin],
  [SB.EXPORT, exportBuiltin],
  [SB.UNSET, unsetBuiltin],
  [SB.LOCAL, localBuiltin],
  [SB.SET, setBuiltin],
  [SB.READ, readBuiltin],
  [SB.MAPFILE, mapfileBuiltin],
  [SB.READARRAY, mapfileBuiltin],
  [SB.SHIFT, shiftBuiltin],
  [SB.GETOPTS, getoptsBuiltin],
  [SB.LET, letBuiltin],
  [SB.TRAP, trapBuiltin],
  [SB.SHOPT, shoptBuiltin],
  [SB.UMASK, umaskBuiltin],
  [SB.ALIAS, aliasBuiltin],
  [SB.UNALIAS, unaliasBuiltin],
  [SB.EXEC, execBuiltin],
  [SB.TEST, testBuiltin],
  [SB.BRACKET, testBuiltin],
  [SB.DOUBLE_BRACKET, testBuiltin],
  [SB.ECHO, echoBuiltin],
  [SB.PRINTF, printfBuiltin],
  [SB.SOURCE, sourceBuiltin],
  [SB.DOT, sourceBuiltin],
  [SB.EVAL, evalBuiltin],
  [SB.COMMAND, commandBuiltin],
  [SB.TYPE, typeBuiltin],
  [SB.WHICH, whichBuiltin],
  [SB.TRUE, trueBuiltin],
  [SB.FALSE, falseBuiltin],
  [SB.COLON, colonBuiltin],
  [SB.BREAK, breakBuiltin],
  [SB.CONTINUE, continueBuiltin],
  [SB.RETURN, returnBuiltin],
  [SB.EXIT, exitBuiltin],
  [SB.PRINTENV, printenvBuiltin],
  [SB.ENV, envBuiltin],
  [SB.WHOAMI, whoamiBuiltin],
  [SB.MAN, manBuiltin],
  [SB.HISTORY, historyBuiltin],
  [SB.SLEEP, sleepBuiltin],
  [SB.BASH, bashBuiltin],
  [SB.SH, bashBuiltin],
  [SB.XARGS, xargsBuiltin],
  [SB.TIMEOUT, timeoutBuiltin],
])
